File: solution/java/46_Permutations.java
class Solution {
    //Time Complexity - O(n!)， Space Complexity - O(n)。
    //backtracking + DFS
    //ims, 100% faster
    
    public List<List<Integer>> permute(int[] nums) {
        
        List<List<Integer>> res = new ArrayList<>();
        getPermute(nums, res, new ArrayList<>());
        return res;
    }
    
    public void getPermute(int [] nums, List<List<Integer>> res, List<Integer> tempList){
        
        if(tempList.size()==nums.length){
            res.add(new ArrayList<>(tempList));
        }else{
        
            for(int i=0;i<nums.length;i++){
                if(tempList.contains(nums[i])) continue; //point

                tempList.add(nums[i]);
                getPermute(nums, res, tempList);
                tempList.remove(tempList.size()-1);
            }
        }
        
    }
    
    //e.g. 1 2 3
    //[1] 2~3
        //[1,2]
            //[1,2,3]
        //[1,3]
            //[1,3,2]
    //[2] 1 3
        //
}

File: solution/cpp/46_Permutations.cpp
class Solution {
public:
    vector<vector<int>> permute(vector<int>& nums) {

        vector<int> list;
        //sol.1 swap, 4ms 75% faster
        findPermute(nums, 0);
        //sol.2, 4ms 75% faster
        // findPermute(nums, list);
        return res;
    }
    
private:
    vector<vector<int>> res;
   
    //sol.1
    void findPermute(vector<int>& nums, int start){
        if(start>=nums.size()){
            res.push_back(vector<int>(nums));
            return;
        }
        
        for(int i=start;i<nums.size();i++){
            swap(nums[start], nums[i]);
		    findPermute(nums, start + 1);
		    swap(nums[start], nums[i]);
        }
        
    }
    
//     //sol.2
//     void findPermute(vector<int>& nums, vector<int>& list){
//         if(nums.size()==0){
//             res.push_back(vector<int>(list));
//             return;
//         }
        
//         for(int i=0;i<nums.size();i++){
//             int cur=nums[i];
//             nums.erase(nums.begin()+i);
//             list.push_back(cur);
                
//             findPermute(nums, list);
            
//             nums.insert(nums.begin()+i,cur);
//             list.pop_back();
//         }
        
//     }
};
